hawkbit-extension-artifact-repository-s3/src/main/java/org/eclipse/hawkbit/artifact/repository/S3Artifact.java

```diff
@@ -37,14 +58,132 @@
+     */
+    public static S3Artifact get(final AmazonS3 amazonS3, final S3RepositoryProperties s3Properties, final String key,
+            final String artifactId) {
+        final S3Object s3Object = getS3ObjectOrThrowException(amazonS3, s3Properties.getBucketName(), key);
+
+        final ObjectMetadata objectMetadata = s3Object.getObjectMetadata();
+        final DbArtifactHash artifactHash = createArtifactHash(artifactId, objectMetadata);
+        return new S3Artifact(s3Object, amazonS3, s3Properties, key, artifactId, artifactHash,
+                objectMetadata.getContentLength(), objectMetadata.getContentType());
+    }
+
+    /**
+     * Create a new instance of {@link S3Artifact}. In this case it is not checked
+     * if an artifact with the given values exists. The S3 object is empty.
+     *
+     * @param amazonS3
+     *            connection to the AmazonS3
+     * @param s3Properties
+     *            used to retrieve the bucket name
+     * @param key
+     *            of the artifact
+     * @param hashes
+     *            instance of {@link DbArtifactHash}
+     * @param size
+     *            of the artifact
+     * @param contentType
+     *            of the artifact
+     * @return an instance of {@link S3Artifact} with an empty {@link S3Object}
+     */
+    public static S3Artifact create(final AmazonS3 amazonS3, final S3RepositoryProperties s3Properties,
+            final String key, final DbArtifactHash hashes, final Long size, final String contentType) {
+        return new S3Artifact(amazonS3, s3Properties, key, hashes.getSha1(), hashes, size, contentType);
+    }
+
+    /**
+     * Verify if the {@link S3Object} exists
+     *
+     * @return result of {@link AmazonS3}#doesObjectExist
+     */
+    public boolean exists() {
+        return amazonS3.doesObjectExist(s3Properties.getBucketName(), key);
     }
 
     @Override
     public String toString() {
         return "S3Artifact [key=" + key + ", getArtifactId()=" + getArtifactId() + ", getHashes()=" + getHashes()
                 + ", getSize()=" + getSize() + ", getContentType()=" + getContentType() + "]";
     }
+
+    @Override
+    public InputStream getFileInputStream() {
+        LOG.debug("Get file input stream for s3 object with key {}", key);
+        refreshS3ObjectIfNeeded();
+        return s3InputStream;
+    }
+    
+    private void refreshS3ObjectIfNeeded() {
+        if (s3Object == null || s3InputStream == null) {
+            LOG.info("Initialize S3Object in bucket {} with key {}", s3Properties.getBucketName(), key);
+            s3Object = amazonS3.getObject(s3Properties.getBucketName(), key);
+            s3InputStream = WrappedS3InputStream.wrap(s3Object.getObjectContent());
+        }
+    }
+
+    private static S3Object getS3ObjectOrThrowException(AmazonS3 amazonS3, String bucketName, String key) {
+        final S3Object s3Object = amazonS3.getObject(bucketName, key);
```

Comment:
we should also catch `SdkClientException` as well as `AmazonServiceException` here